main.go

```diff
@@ -123,3 +153,242 @@ func Handler(rw http.ResponseWriter, req *http.Request) {
+	}
+
+	// 3: send .finished event
+	return sendGetSLIFinishedEvent(event, eventData, sliResults, nil, keptnCtx)
+}
+
+func retrieveMetrics(event cloudevents.Event, eventData *keptnv2.GetSLITriggeredEventData, log keptncommon.LoggerInterface) ([]*keptnv2.SLIResult, error) {
+	log.Info("Retrieving Prometheus metrics")
+
+	clusterConfig, err := rest.InClusterConfig()
+	if err != nil {
+		log.Error("could not create Kubernetes cluster config")
+		return nil, errors.New("could not create Kubernetes client")
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		log.Error("could not create Kubernetes client")
+		return nil, errors.New("could not create Kubernetes client")
+	}
+
+	prometheusApiURL, err := getPrometheusApiURL(eventData.Project, kubeClient.CoreV1(), log)
+	if err != nil {
+		return nil, err
+	}
+
+	eventBrokerURL := os.Getenv(eventbroker)
+	if eventBrokerURL == "" {
+		eventBrokerURL = "http://event-broker/keptn"
+	}
+
+	keptnHandler, err := keptnv2.NewKeptn(&event, keptncommon.KeptnOpts{EventBrokerURL: eventBrokerURL})
+	if err != nil {
+		return nil, err
+	}
+
+	prometheusHandler := utils.NewPrometheusHandler(prometheusApiURL, eventData.Project, eventData.Stage, eventData.Service, eventData.GetSLI.CustomFilters)
+
+	projectCustomQueries, err := getCustomQueries(keptnHandler, eventData.Project, eventData.Stage, eventData.Service, log)
+	if err != nil {
+		log.Error("Failed to get custom queries for project " + eventData.Project)
+		log.Error(err.Error())
+		return nil, err
+	}
+
+	if projectCustomQueries != nil {
+		prometheusHandler.CustomQueries = projectCustomQueries
+	}
+
+	var sliResults []*keptnv2.SLIResult
+
+	for _, indicator := range eventData.GetSLI.Indicators {
+		log.Info("Fetching indicator: " + indicator)
+		sliValue, err := prometheusHandler.GetSLIValue(indicator, eventData.GetSLI.Start, eventData.GetSLI.End, log)
+		if err != nil {
+			sliResults = append(sliResults, &keptnv2.SLIResult{
+				Metric:  indicator,
+				Value:   0,
+				Success: false,
+				Message: err.Error(),
+			})
+		} else if math.IsNaN(sliValue) {
+			sliResults = append(sliResults, &keptnv2.SLIResult{
+				Metric:  indicator,
+				Value:   0,
+				Success: false,
+				Message: "SLI value is NaN",
+			})
+		} else {
+			sliResults = append(sliResults, &keptnv2.SLIResult{
+				Metric:  indicator,
+				Value:   sliValue,
+				Success: true,
+			})
+		}
+	}
+	return sliResults, nil
+}
+
+func getCustomQueries(keptnHandler *keptnv2.Keptn, project string, stage string, service string, logger keptncommon.LoggerInterface) (map[string]string, error) {
+	logger.Info("Checking for custom SLI queries")
+
+	customQueries, err := keptnHandler.GetSLIConfiguration(project, stage, service, sliResourceURI)
+	if err != nil {
+		return nil, err
+	}
+
+	return customQueries, nil
+}
+
+func getPrometheusApiURL(project string, kubeClient v1.CoreV1Interface, logger keptncommon.LoggerInterface) (string, error) {
```

Comment:
Do we still need this function if we already have the `prometheusEndpoint` configured via an environment variable?

Reply:
@jetzlstorfer what do you think - should we take the approach of configuring the prometheus endpoint as an environment variable, or reading it from the K8s secret using the K8s library? Currently we use one approach for the SLI service features (reading it via the k8s library), while the other features of the prometheus-service rely on having the prometheus endpoint as an environment variable. I think this should be streamlined to avoid confusion

Reply:
yes this should be definitely streamlined. for me both options are fine.
most important here is that the approach is streamlined with other integrations such as the Dynatrace service and Dynatrace SLI service or Neoload SLI service.
What is your recommendation here?

Reply:
for the dynatrace service we currently use the approach of going with the K8s library and reading the secret on demand. This provides the benefit of being able to update the secrets containing the credentials at runtime without the need to restart the pod of the service referencing those secret. Also, this provides more flexibility because this way, different credentials can be defined for different projects. The downside of this approach is that it requires more implementation effort to retrieve the secret via the K8s library - but in my opinion the benefits outweigh this disadvantage in that case

Reply:
Thanks, @bacherfl @jetzlstorfer for the review. As per the current implementationation, First, it will look into the k8s secrets, if it is not set then it will the PROMETHEUS_ENDPOINT from the deployment env.

Reply:
Thanks for the explanation, I think that is a reasonable approach